### PR TITLE
Jl/transaction multichain nonce lock by chain

### PR DESCRIPTION
**Pull request by jiexi**
## Explanation

Naively couples existing mutex with nonceLock mutex. Doesn't seem to cause a deadlock

### `@metamask/package-a`

- **<CATEGORY>**: Your change here
- **<CATEGORY>**: Your change here

### `@metamask/package-b`

- **<CATEGORY>**: Your change here
- **<CATEGORY>**: Your change here

## Checklist

- [ ] I've updated the test suite for new or updated code as appropriate
- [ ] I've updated documentation (JSDoc, Markdown, etc.) for new or updated code as appropriate
- [ ] I've highlighted breaking changes using the "BREAKING" category above as appropriate

### CODE REVIEW

**Comment by adonesky1** [comment]
I don't think this is what we want to do. I mean currently since there is a single nonceTracker its just redundant but assuming you're suggesting that we only allow one nonce to be issued at a time even when we multiplex this I don't think that's the right approach. We will want to enable concurrent transaction submissions as long as their on different chains. But perhaps I'm misreading your angle here.

**Review by adonesky1 — COMMENTED** [review]
I don't think this is what we want to do. I mean currently since there is a single nonceTracker its just redundant but assuming you're suggesting that we only allow one nonce to be issued at a time even when we multiplex this I don't think that's the right approach. We will want to enable concurrent transaction submissions as long as their on different chains. But perhaps I'm misreading your angle here.

**Comment by jiexi** [comment]
Your read into my approach is correct. If we want concurrent nonceLocks, then we will need to keep a map of chainId to mutex. I'll update this PR to reflect that

File: packages/transaction-controller/src/TransactionController.ts [hunk 1]
```diff
@@ -1142,6 +1144,11 @@ export class TransactionController extends BaseControllerV1<
 
   startTrackingByNetworkClientId(networkClientId: NetworkClientId) {
     const networkClient = this.getNetworkClientById(networkClientId);
+    const chainId = networkClient.configuration.chainId;
+    if(!this.nonceMutexByChainId.get(chainId)) {
+      this.nonceMutexByChainId.set(chainId, new Mutex())
+    }
```
Comment (jiexi) [inline packages/transaction-controller/src/TransactionController.ts:1150]: need to clear mutexes on network removal either keeping count of networks left with chainId, or just checking if any remaining networks with chainId

**Comment by jiexi** [comment]
uses nonces keyed by chainId now

File: packages/transaction-controller/src/TransactionController.ts [hunk 2]
```diff
@@ -1585,11 +1596,35 @@ export class TransactionController extends BaseControllerV1<
...
+    if (networkClientId) {
+      const networkClient = this.getNetworkClientById(networkClientId);
+      nonceMutexForChainId = this.nonceMutexByChainId.get(
+        networkClient.configuration.chainId,
+      );
+    }
+
+    const releaseLockForChainId = await nonceMutexForChainId?.acquire();
+    try {
+      const nonceLock = await this.nonceTracker.getNonceLock(address);
```
Comment (adonesky1) [inline packages/transaction-controller/src/TransactionController.ts:1616]: I believe we'll want to get the nonceLock from the correct nonceTracker for the networkClientId here. I realize we haven't yet instantiated the trackingSet so this isn't possible yet.

**Review by adonesky1 — APPROVED** [review]

**Pull request merged**